aten/src/THCUNN/generic/TemporalUpSamplingNearest.cu

```diff
@@ -5,152 +5,91 @@
-     THCTensor_(resize3d)(state, output,
-                          THCTensor_(size)(state, input, 0),
-                          THCTensor_(size)(state, input, 1),
-                          outputWidth);
-  }
-
-  input = THCTensor_(newContiguous)(state, input);
-  // This is for allocating output Tensor
-  int64_t no_elements = 1;
-  for(int i = 0; i < input->_dim(); i++){
-    no_elements *= input->size[i];
-  }
-  no_elements *= scale_factor;
-
-  int d1;
-  int d2;
+  int nbatch = THCTensor_(size)(state, input, 0);
+  int channels = THCTensor_(size)(state, input, 1);
+  int inputWidth  = THCTensor_(size)(state, input, 2);
 
-  if (input->_dim() == 2) {
-    d1 = output->size[0];
-    d2 = output->size[1];
-  } else {
-    d1 = output->size[1];
-    d2 = output->size[2];
-  }
+  THNN_(TemporalUpSamplingNearest_shapeCheck)(state, input, NULL, nbatch, channels, inputWidth, outputWidth);
+  THAssert(inputWidth > 0 && outputWidth > 0);
 
-  real *input_data = THCTensor_(data)(state, input);
-  real *output_data = THCTensor_(data)(state, output);
-
-  // cuda blocks & threads:
-  int64_t nthreads = 256;
-  // Max number of blocks: http://en.wikipedia.org/wiki/CUDA
-  // 65535 for SM 2.x, 2^32 -1 for >= 3.0
-  // TODO: When we move to SM 3.5 we should update this
-  int64_t n_xblocks = min(max((int)ceil((float)no_elements / nthreads), 1), 65535);
-  int64_t n_yblocks = (int64_t)ceil((float)no_elements / (float)(n_xblocks * nthreads));
-  if (n_yblocks > 65535) {
-    THError("Input size is too large!  aborting");
-  }
-  dim3 blocks(n_xblocks, n_yblocks);
-  dim3 threads(nthreads);
+  THCTensor_(resize3d)(state, output,
+                       THCTensor_(size)(state, input, 0),
+                       THCTensor_(size)(state, input, 1),
+                       outputWidth);
+  THCTensor_(zero)(state, output);
 
-  // kernel:
-  upscale<<<blocks, threads, 0, THCState_getCurrentStream(state)>>> (input_data, output_data, no_elements, scale_factor, d1, d2);
+  input = THCTensor_(newContiguous)(state, input);
+  THCDeviceTensor<real, 3> idata = toDeviceTensor<real, 3>(state, input);
+  THCDeviceTensor<real, 3> odata = toDeviceTensor<real, 3>(state, output);
+
+  const int num_kernels = outputWidth;
+  const int num_threads = THCState_getCurrentDeviceProperties(state)->maxThreadsPerBlock;
+  cudaStream_t stream = THCState_getCurrentStream(state);
+  nearest_neighbor_interp2_kernel<real, accreal> <<<THCCeilDiv(num_kernels, num_threads), num_threads,
+	 0, stream>>>(num_kernels, idata, odata);
   THCudaCheck(cudaGetLastError());
-
-  // final cut:
   THCTensor_(free)(state, input);
 }
 
+
+
 void THNN_(TemporalUpSamplingNearest_updateGradInput)(
            THCState *state,
-           THCTensor *input,
            THCTensor *gradOutput,
            THCTensor *gradInput,
-           int scale_factor)
+           int nbatch,
+	   int nchannels,
+	   int inputWidth,
+	   int outputWidth)
 {
-
+  gradInput = THCTensor_(newContiguous)(state, gradInput);
```

Comment:
Generally you can assume gradInput is contiguous.

Reply:
Actually all these lines making it contiguous was added in my last commit. There was a segfault in every cuda backward functions if without these lines.

Reply:
Hmm that is weird. But they were working before your patch, right?

Reply:
Aha I think I get why it's different. So here is context: with the old kernels, it works without initializing gradInput as contiguous. But with the current kernels(like *linear), it segfaults when gradInput is not initialized as contiguous(without last commit). 
And here are my 2 cents: with old code, it didn't initialize as contiguous but get a real * pointer and pass along to the kernel code. In the kernel we actually use the memory contiguously. https://github.com/pytorch/pytorch/blob/master/aten/src/THCUNN/TemporalUpSamplingNearest.cu#L40
With the new code, we are passing along a THCDeviceTensor, so if it's not initialized as contiguous, it might access some wild memory space during the assignment here. https://github.com/pytorch/pytorch/blob/master/aten/src/THCUNN/TemporalUpSamplingLinear.cu#L72 
Basically we made sure the memory is contiguous in both cases. And I would prefer to keep it as it is since it's consistent with how *linear methods were implemented. 
What do you think?

Reply:
I'm not sure the exact reason for segfault. But I realized that you can't call `newContiguous` here as you are overwriting `gradInput` so (if this line actually makes a new tensor) the output is never written to the input of this function!